jobbergate-api/jobbergate_api/file_validation.py

```diff
@@ -0,0 +1,265 @@
+    """
+    logger.debug("Checking uploaded files, validating source code")
+
+    with UploadedFilesValidationError.check_expressions("Invalid syntax on uploaded file(s)") as check:
+        for f in file_list:
+            suffix = get_suffix(f)
+            if suffix not in syntax_validation_dispatch:
+                continue
+            check(syntax_validation_dispatch[suffix](f.file.read()), f.filename)
+            f.file.seek(0)
+
+
+@register_check_uploaded_files
+def check_uploaded_files_yaml_is_parsable(file_list: List[UploadFile]) -> None:
+    """
+    Check the list of uploaded files to verify if the yaml file agrees with the expected schema.
+
+    Raise UploadedFilesValidationError if the business rules are not met.
+
+    :param List[UploadFile] file_list: Upload file list.
+    """
+    logger.debug("Checking uploaded files, parsing yaml file")
+    with UploadedFilesValidationError.handle_errors(
+        (
+            "Not possible to get the configuration from the yaml file."
+            " Please, verify is all the required fields were provided."
+        )
+    ):
+        for file in file_list:
+            if file.filename.endswith(".yaml"):
+                config = safe_load(file.file.read())
+                file.file.seek(0)
+                ApplicationConfig(**config)
+
+
+SyntaxValidationEquation = Callable[[Union[str, bytes]], bool]
+"""Type alias describing the function signature used to validate file syntax."""
+
+syntax_validation_dispatch: Dict[str, SyntaxValidationEquation] = {}
+"""Dictionary mapping file extensions to the function used to validate their syntax."""
+
+
+def register_syntax_validator(*file_extensions: str):
+    """
+    Use this decorator to register file syntax validation functions.
+
+    It creates a new entry on ``validation_dispatch``, mapping the equation to
+    the file extensions that are provided as arguments.
+
+    Raise ValueError if the provided file extensions do not start with a dot.
+
+    :return ValidationEquation: The decorated function.
+    """
+
+    def decorator(validator):
+        for extension in file_extensions:
+            require_condition(
+                extension.startswith("."),
+                f"File extensions should start with a dot, got the value: {extension}",
+                raise_exc_class=ValueError,
+            )
+            syntax_validation_dispatch[extension] = validator
+
+        @wraps(validator)
+        def wrapper(source_code):
+            return validator(source_code)
+
+        return wrapper
+
+    return decorator
+
+
+@register_syntax_validator(".py")
+def is_valid_python_file(source_code: Union[str, bytes]) -> bool:
+    """
+    Check if a given Python source code is valid by parsing it into an AST node.
+
+    :param Union[str, bytes] source_code: Python source code.
+    :return bool: Boolean indicating if the source code is valid or not.
+    """
+    try:
+        parse(source_code)
+    except SyntaxError:
```

Comment:
It would propagate the entire way back until FastAPI catches it. Should we catch any error on this? That is a good point, at least the error handles would give a nice message.

Reply:
Yes, I was just thinking if another error came out of `parse`, it would also indicate something wrong with the source file. Maybe `parse` only gives `SyntaxError`, but I think there's no need to be that restrictive.

Reply:
solved in [3617cbb](https://github.com/omnivector-solutions/jobbergate/pull/128/commits/3617cbbe4fdf187c02b987ac6b101b3d6020119d).